Source/WebKit/UIProcess/Automation/WebDriverBidiProcessor.cpp

```diff
@@ -111,19 +117,183 @@ void WebDriverBidiProcessor::sendMessageToFrontend(const String& message)
+
+    auto presentationHint = createType == Inspector::Protocol::BidiBrowsingContext::CreateType::Tab ? Inspector::Protocol::Automation::BrowsingContextPresentation::Tab : Inspector::Protocol::Automation::BrowsingContextPresentation::Window;
+    session->createBrowsingContext(presentationHint, [callback = WTFMove(callback)](CommandResultOf<BrowsingContext, Inspector::Protocol::Automation::BrowsingContextPresentation>&& result) {
+        if (!result) {
+            callback(makeUnexpected(result.error()));
+            return;
+        }
+
+        auto [resultContext, resultPresentation] = WTFMove(result.value());
+        callback(resultContext);
+    });
+}
+
+void WebDriverBidiProcessor::getTree(const BrowsingContext& optionalRoot, std::optional<double>&& optionalMaxDepth, CommandCallback<Ref<JSON::ArrayOf<Protocol::BidiBrowsingContext::Info>>>&& callback)
+{
+    RefPtr session = m_session.get();
+    if (!session)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+
+    // FIXME: implement `root` option.
+    // FIXME: implement `maxDepth` option.
+
+    Vector<Ref<WebPageProxy>> pages;
+    for (Ref process : session->protectedProcessPool()->processes()) {
+        for (Ref page : process->pages()) {
+            if (!page->isControlledByAutomation())
+                continue;
+
+            pages.append(WTFMove(page));
+
+            // FIXME: add browsingContexts for frame tree nodes.
+        }
+    }
+
+    Ref<JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>> infos = JSON::ArrayOf<Inspector::Protocol::BidiBrowsingContext::Info>::create();
+    for (Ref page : pages) {
+        // FIXME: implement `parent` field.
+        // FIXME: implement `children` field.
+        // FIXME: implement `originalOpener` field.
+        // FIXME: implement `clientWindow` field.
+        // FIXME: implement `userContext` field.
+
+        infos->addItem(Inspector::Protocol::BidiBrowsingContext::Info::create()
+            .setContext(session->handleForWebPageProxy(page))
+            .setUrl(page->currentURL())
+            .setClientWindow("placeholder_window"_s)
+            .setUserContext("placeholder_context"_s)
+            .release());
+    }
+
+    callback({ { WTFMove(infos) } });
+}
+
+// https://www.w3.org/TR/webdriver/#dfn-session-page-load-timeout
+static const Seconds defaultPageLoadTimeout = 300_s;
+
+void WebDriverBidiProcessor::navigate(const BrowsingContext& browsingContext, const String& url, std::optional<Inspector::Protocol::BidiBrowsingContext::ReadinessState>&& optionalReadinessState, CommandCallbackOf<String, Inspector::Protocol::BidiBrowsingContext::Navigation>&& callback)
+{
+    RefPtr session = m_session.get();
+    if (!session)
+        ASYNC_FAIL_WITH_PREDEFINED_ERROR(InternalError);
+
+    PageLoadStrategy strategy = PageLoadStrategy::None;
+    switch (optionalReadinessState.value_or(Inspector::Protocol::BidiBrowsingContext::ReadinessState::None)) {
```

Comment:
can we make this into a `static` function so that we can avoid writing it again below?